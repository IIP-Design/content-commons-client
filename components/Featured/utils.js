import { capitalizeFirst } from 'lib/utils';
import { normalizeItem } from 'lib/elastic/parser';
import { postTypeAggRequest } from 'lib/elastic/api';

/**
 * Get a string of up to three categories from a content item
 * @param {Object} item Normalized data for a give content item
 * @returns {string} Lowercase list of category names separated by ·
 */
export const getCategories = item => {
  const cats = item?.categories?.slice( 0, 3 ).reduce( ( acc, cat, index, arr ) => {
    const c = acc + cat.name.toLowerCase();

    return index < arr.length - 1 && index < 2 ? `${c} · ` : c;
  }, '' );

  return cats;
};

/**
 * Updates the the featured context
 * @param {Promise[]} array Group of promises that resolve the recents properties
 * @param {dispatch} dispatch A reducer dispatch function
 */
export const getFeatured = async ( array, dispatch ) => {
  const resArr = await Promise.all( array ).catch( err => {
    dispatch( {
      type: 'LOAD_FEATURED_FAILED',
    } );
  } );

  console.dir( resArr );

  const priorities = {};
  const recents = {};
  let items;

  if ( resArr ) {
    resArr.forEach( res => {
      if ( res?.data?.hits?.hits ) {
        items = res.data.hits.hits.map( item => normalizeItem( item, res.locale ) );

        switch ( res.component ) {
          case 'priorities':
            priorities[res.term] = items;
            break;
          case 'packages':
          case 'recents':
            recents[res.postType] = items;
            break;
          default:
            break;
        }
      }
    } );

    dispatch( {
      type: 'LOAD_FEATURED_SUCCESS',
      payload: {
        priorities,
        recents,
      },
    } );
  }
};


/**
 *
 * @param {dispatch} dispatch A reducer dispatch function
 * @param {Object} user User date object
 */
export const loadPostTypes = async ( dispatch, user ) => {
  dispatch( { type: 'LOAD_POST_TYPES_PENDING' } );

  let response;

  try {
    response = await postTypeAggRequest( user );
  } catch ( err ) {
    dispatch( { type: 'LOAD_POST_TYPES_FAILED' } );

    return;
  }

  if ( response?.aggregations?.postType?.buckets ) {
    const { buckets } = response.aggregations.postType;

    const filtered = buckets.filter( type => type.key !== 'courses' && type.key !== 'page' && type.key !== 'package' );

    const payload = filtered.map( type => {
      let displayName = capitalizeFirst( type.key );

      if ( type.key === 'post' ) displayName = 'Article';
      if ( type.key === 'document' ) displayName = 'Press Releases and Guidance';

      return {
        key: type.key,
        display_name: displayName,
        count: type.doc_count,
      };
    } );

    dispatch( {
      type: 'LOAD_POST_TYPES_SUCCESS',
      payload,
    } );
  } else {
    dispatch( {
      type: 'LOAD_POST_TYPES_SUCCESS',
      payload: [],
    } );
  }
};
