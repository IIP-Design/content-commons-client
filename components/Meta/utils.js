/**
 * Check if content exists for a given meta property
 * If not, filters it from the meta object
 *
 * @param {Object} meta All possible meta tags
 * @returns {Object} Only those meta tags for which there is content
 */
const filterEmpty = meta => {
  const filteredMeta = {};
  const metaArr = Object.entries( meta );

  metaArr.forEach( item => {
    if ( item[1] ) {
      const property = item[0];
      const value = item[1];

      filteredMeta[property] = value;
    }
  } );

  return filteredMeta;
};

/**
 * Converts the content item object into a metavalues array using the checkAndPush method
 *
 * @param {object} item Object of metadata for a given content item
 * @param {string} url The URL of the page to be rendered
 * @returns {array} Array of { property: content } key pairs
 */
export const getOgTags = ( item, url ) => {
  const {
    selectedLanguageUnit,
    title,
    type,
  } = item;

  const isVideo = type && type === 'video';

  /* TEMP vars for Document use case */
  const trimContent = item.content && item.content.rawText && `${item.content.rawText.substring( 0, 152 )}...`;
  const description = item.description || trimContent;

  const videoSrc = selectedLanguageUnit && selectedLanguageUnit.source && selectedLanguageUnit.source[0]
    ? selectedLanguageUnit.source[0]
    : {};
  const streamSrc = videoSrc.stream && videoSrc.stream.link ? videoSrc.stream.link : '';

  const meta = {
    'og:type': isVideo ? 'video.other' : 'article',
    'og:url': url,
    'twitter:card': isVideo ? 'player' : 'summary',
    'og:description': description,
    'og:title': title,
    'og:video:url': isVideo ? streamSrc : '',
    'twitter:player': isVideo ? streamSrc : '',
    'twitter:player:height': isVideo ? '196px' : '',
    'twitter:player:width': isVideo ? '350px' : '',
  };

  return filterEmpty( meta );
};
