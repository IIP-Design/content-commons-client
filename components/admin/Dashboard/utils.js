import { getProjectsType } from 'lib/graphql/util';
import {
  DELETE_GRAPHIC_PROJECT_MUTATION,
  TEAM_GRAPHIC_PROJECTS_QUERY,
  TEAM_GRAPHIC_PROJECTS_COUNT_QUERY,
  GRAPHIC_PROJECT_IMAGE_FILES_QUERY,
  UNPUBLISH_GRAPHIC_PROJECT_MUTATION,
  UPDATE_GRAPHIC_STATUS_MUTATION,
  GRAPHIC_PROJECTS_META_QUERY,
} from 'lib/graphql/queries/graphic';
import {
  DELETE_PACKAGE_MUTATION,
  PACKAGE_FILES_QUERY,
  TEAM_PACKAGES_COUNT_QUERY,
  TEAM_PACKAGES_QUERY,
  UNPUBLISH_PACKAGE_MUTATION,
  UPDATE_PACKAGE_STATUS_MUTATION,
  PACKAGES_META_QUERY,
} from 'lib/graphql/queries/package';
import {
  DELETE_VIDEO_PROJECT_MUTATION,
  TEAM_VIDEO_PROJECTS_COUNT_QUERY,
  TEAM_VIDEO_PROJECTS_QUERY,
  UNPUBLISH_VIDEO_PROJECT_MUTATION,
  UPDATE_VIDEO_STATUS_MUTATION,
  VIDEO_PROJECT_FILES_QUERY,
  VIDEO_PROJECTS_META_QUERY,
} from 'lib/graphql/queries/video';
import { TEAM_PLAYBOOKS_QUERY, TEAM_PLAYBOOKS_COUNT_QUERY } from 'lib/graphql/queries/playbook';

/**
 * Returns the project type based off of the content type available to a team
 *
 * @param {Object} team team data received from GraphQL
 */
export const getContentTypesFromTeam = team => {
  const contentTypes = team?.contentTypes ? team.contentTypes : '';

  const type = getProjectsType( contentTypes );

  return type || null;
};

/**
 * Returns the expected queries depending on what content types a team has access to
 *
 * @param {Object} team team data received from GraphQL
 * @returns {Object} list of relevant queries
 */
export const setQueries = team => {
  const queries = {};

  switch ( getContentTypesFromTeam( team ) ) {
    case 'graphicProjects':
      queries.content = TEAM_GRAPHIC_PROJECTS_QUERY;
      queries.count = TEAM_GRAPHIC_PROJECTS_COUNT_QUERY;
      queries.files = GRAPHIC_PROJECT_IMAGE_FILES_QUERY;
      queries.metaContent = GRAPHIC_PROJECTS_META_QUERY;
      queries.remove = DELETE_GRAPHIC_PROJECT_MUTATION;
      queries.status = UPDATE_GRAPHIC_STATUS_MUTATION;
      queries.unpublish = UNPUBLISH_GRAPHIC_PROJECT_MUTATION;

      return queries;
    case 'videoProjects':
      queries.content = TEAM_VIDEO_PROJECTS_QUERY;
      queries.count = TEAM_VIDEO_PROJECTS_COUNT_QUERY;
      queries.files = VIDEO_PROJECT_FILES_QUERY;
      queries.metaContent = VIDEO_PROJECTS_META_QUERY;
      queries.remove = DELETE_VIDEO_PROJECT_MUTATION;
      queries.status = UPDATE_VIDEO_STATUS_MUTATION;
      queries.unpublish = UNPUBLISH_VIDEO_PROJECT_MUTATION;

      return queries;
    case 'packages':
      queries.content = TEAM_PACKAGES_QUERY;
      queries.count = TEAM_PACKAGES_COUNT_QUERY;
      queries.files = PACKAGE_FILES_QUERY;
      queries.metaContent = PACKAGES_META_QUERY;
      queries.remove = DELETE_PACKAGE_MUTATION;
      queries.status = UPDATE_PACKAGE_STATUS_MUTATION;
      queries.unpublish = UNPUBLISH_PACKAGE_MUTATION;

      return queries;

    case 'playbooks':
      queries.content = TEAM_PLAYBOOKS_QUERY;
      queries.count = TEAM_PLAYBOOKS_COUNT_QUERY;
      // The following queries are placeholders and for the wrong content type
      // As such, these operations will not work properly
      queries.files = PACKAGE_FILES_QUERY;
      queries.metaContent = PACKAGES_META_QUERY;
      queries.remove = DELETE_PACKAGE_MUTATION;
      queries.status = UPDATE_PACKAGE_STATUS_MUTATION;
      queries.unpublish = UNPUBLISH_PACKAGE_MUTATION;

      return queries;
    default:
      return queries;
  }
};
