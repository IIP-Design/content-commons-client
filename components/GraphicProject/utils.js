import { getTransformedLangTaxArray, getCount } from 'lib/utils';

const structureLangObj = item => ( {
  language: {
    language_code: item.language.languageCode,
    text_direction: item.language.textDirection,
    locale: item.language.locale,
    display_name: item.language.displayName,
    native_name: item.language.nativeName,
  },
} );

export const normalizeGraphicProjectByAPI = ( { file, useGraphQl = false } ) => {
  const graphicObj = {
    id: file.id || '',
    type: 'graphic',
    site: file.site || '',
    title: file.title || '',
    projectType: file.projectType || '',
    published: file.published || '',
    modified: file.modified || '',
    owner: file.owner || '',
    desc: file.desc || '',
    descInternal: file.descInternal || '',
    copyright: file.copyright || '',
    images: file.images || [],
    supportFiles: file.supportFiles || [],
    categories: file.categories || '',
  };

  // Elastic API
  if ( !useGraphQl ) {
    const imagesWithFileNameProp = file.images.map( img => ( {
      ...img,
      filename: img.filename,
    } ) );

    const supportFilesWithFileNameProp = Array.isArray( file.supportFiles ) && file.supportFiles.map( supFile => {
      const filenameFromSrc = supFile.url.slice( supFile.url.lastIndexOf( '/' ) + 1 );

      return {
        ...supFile,
        filename: filenameFromSrc,
      };
    } );

    const esObj = {
      images: imagesWithFileNameProp || [],
      supportFiles: supportFilesWithFileNameProp || [],
    };

    return { ...graphicObj, ...esObj };
  }

  // GraphQL API
  if ( useGraphQl ) {
    const structuredImages = file.images.map( img => {
      const { dimensions } = img;

      return {
        ...img,
        ...structureLangObj( img ),
        url: img.signedUrl,
        width: dimensions.width,
        height: dimensions.height,
      };
    } );

    const structuredSupportFiles = Array.isArray( file.supportFiles ) && file.supportFiles.map( supFile => ( {
      ...supFile,
      ...structureLangObj( supFile ),
      srcUrl: supFile.signedUrl,
    } ) );

    const gqlObj = {
      alt: file.alt || '',
      projectType: file.type || '',
      published: file.publishedAt || '',
      modified: file.updatedAt || '',
      owner: file.team?.name || '',
      desc: file.descPublic || '',
      images: structuredImages || [],
      supportFiles: structuredSupportFiles || [],
      categories: getTransformedLangTaxArray( file.categories ) || [],
    };

    return { ...graphicObj, ...gqlObj };
  }

  return graphicObj;
};

// Use Twitter graphics as default for display otherwise whatever graphic image is available
export const filterGraphicImgs = images => {
  const containsTwitterImgs = images.some( img => img.social.includes( 'Twitter' ) );

  if ( containsTwitterImgs ) {
    return images.filter( img => img.social.includes( 'Twitter' ) );
  }

  return images;
};

export const getGraphicImgsBySocial = ( images, platform, useGraphQl = false ) => {
  if ( !getCount( images ) ) return [];

  const filteredImgs = images.reduce( ( allImgs, img ) => {
    let platformImgs = [];
    let condition = img.social.includes( platform );

    if ( useGraphQl ) {
      platformImgs = img.social.filter( s => s.name === platform );
      condition = getCount( platformImgs );
    }

    if ( condition ) {
      allImgs.push( img );
    }

    return allImgs;
  }, [] );

  return getCount( filteredImgs ) ? filteredImgs : images;
};
