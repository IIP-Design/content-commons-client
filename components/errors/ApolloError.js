import React from 'react';
import PropTypes from 'prop-types';
import { getApolloErrors } from 'lib/utils';


const ApolloError = props => {
  const buttonStyle = {
    color: 'red',
    fontSize: '14px',
    marginTop: '.5rem',
    marginBottom: '1rem',
  };

  const compileErrors = () => {
    const { error } = props;

    let errs = [];

    if ( error ) {
      errs = getApolloErrors( error );
    }

    if ( errs[0] === 'This token is either invalid or expired!' ) {
      return (
        <div>
          { 'Your link has expired. Please return to the ' }
          <a href="/login">login page</a>
          { /**/ }
          .
        </div>
      );
    }

    return errs.join( '\n' ).replace( 'AuthenticationError:', '' )
      .trim();
  };


  return (
    <div style={ buttonStyle }>
      { compileErrors() }
    </div>
  );
};

ApolloError.propTypes = {
  error: PropTypes.object,
};

export default ApolloError;
