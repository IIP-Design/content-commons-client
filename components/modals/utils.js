import GraphicProject from 'components/GraphicProject/GraphicProject';
import Post from 'components/Post/Post';
import Video from 'components/Video/Video';
import Document from 'components/Document/Document';
/**
 * Check content type of an item and passes in the correct component to the modal
 * @param {Object} item data from the selected item
 */
export const getModalContent = item => {
  const noContent = <div>No content currently available</div>;

  if ( item ) {
    switch ( item.type ) {
      case 'graphic':
        return <GraphicProject item={ item } displayAsModal />;

      case 'video':
        return <Video item={ item } />;

      case 'post':
        return <Post item={ item } />;

      case 'document':
        return <Document item={ item } displayAsModal />;

      default:
        return noContent;
    }
  }

  return noContent;
};
