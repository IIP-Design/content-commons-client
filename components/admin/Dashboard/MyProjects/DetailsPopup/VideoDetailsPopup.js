import React from 'react';
import PropTypes from 'prop-types';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import ApolloError from 'components/errors/ApolloError';
import { formatBytes, getCount, getS3Url } from 'lib/utils';

const VIDEO_PROJECT_FILES_QUERY = gql`
  query VideoProjectFiles( $id: ID! ) {
    videoProject( id: $id ) {
      id
      projectType
      units {
        id        
        files {
          id
          filesize
          quality
          url
          language {
            id
            displayName
          }
          use {
            id
            name
          }
        }
      }
      supportFiles(
        where: {
          filetype: "srt"
        }
      ) {
        id
        url
        filesize
        language {
          id
          displayName
        }        
      }
    }
  }
`;

const getValidFiles = files => (
  files.filter( file => file )
);

const getVideoFiles = units => (
  units.reduce( ( acc, unit ) => {
    if ( !unit.files ) return;
    const validFiles = getValidFiles( unit.files );

    return [...acc, ...validFiles];
  }, [] )
);

const VideoDetailsPopup = props => (
  <Query query={ VIDEO_PROJECT_FILES_QUERY } variables={ { id: props.id } }>
    {
      ( { loading, error, data } ) => {
        if ( loading ) return <p>Loading....</p>;
        if ( error ) return <ApolloError error={ error } />;
        if ( !data.videoProject ) return null;

        const {
          units,
          supportFiles,
        } = data.videoProject;

        const videoFiles = getVideoFiles( units );

        if ( getCount( videoFiles ) || getCount( supportFiles ) ) {
          return (
            <div className="details-files">
              <ul>
                { videoFiles && videoFiles.map( vidFile => {
                  if ( !vidFile || getCount( vidFile ) === 0 ) return null;

                  return (
                    <li key={ vidFile.id }>
                      { vidFile.use && vidFile.use.name }
                      { ' | ' }
                      <a href={ getS3Url( vidFile.url ) }>{ vidFile.quality }</a>
                      { ' | ' }
                      <a href={ getS3Url( vidFile.url ) }>
                        { vidFile.language && vidFile.language.displayName }
                        { ' ' }
                        (
                        { formatBytes( vidFile.filesize ) }
                        )
                      </a>
                    </li>
                  );
                } ) }
                { supportFiles && supportFiles.map( sprtFile => {
                  if ( !sprtFile || getCount( sprtFile ) === 0 ) return null;

                  return (
                    <li key={ sprtFile.id }>
                      { 'SRT | ' }
                      <a href={ getS3Url( sprtFile.url ) }>{ sprtFile.language && sprtFile.language.displayName }</a>
                      { ' | ' }
                      <a href={ getS3Url( sprtFile.url ) }>{ formatBytes( sprtFile.filesize ) }</a>
                    </li>
                  );
                } ) }
              </ul>
            </div>
          );
        }

        return <p>There are no supporting video files.</p>;
      }
    }
  </Query>
);

VideoDetailsPopup.propTypes = {
  id: PropTypes.string,
};

export default VideoDetailsPopup;

export { getValidFiles, getVideoFiles, VIDEO_PROJECT_FILES_QUERY };
