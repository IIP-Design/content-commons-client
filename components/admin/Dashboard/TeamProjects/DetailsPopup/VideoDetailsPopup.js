import React from 'react';
import PropTypes from 'prop-types';
import { gql, useQuery } from '@apollo/client';

import ApolloError from 'components/errors/ApolloError';
import { formatBytes, getCount } from 'lib/utils';

const VIDEO_PROJECT_FILES_QUERY = gql`
  query VideoProjectFiles( $id: ID! ) {
    videoProject( id: $id ) {
      id
      projectType
      units {
        id        
        files {
          id
          filesize
          quality
          url
          language {
            id
            displayName
          }
          use {
            id
            name
          }
        }
      }
      supportFiles(
        where: {
          filetype: "srt"
        }
      ) {
        id
        url
        filesize
        language {
          id
          displayName
        }        
      }
    }
  }
`;

const getValidFiles = files => (
  files.filter( file => file && getCount( file ) > 0 )
);

const getVideoFiles = units => (
  units.reduce( ( acc, unit ) => {
    if ( !unit.files ) return;
    const validFiles = getValidFiles( unit.files );

    return [...acc, ...validFiles];
  }, [] )
);

const VideoDetailsPopup = ( { id } ) => {
  const { loading, error, data } = useQuery( VIDEO_PROJECT_FILES_QUERY, {
    variables: { id },
  } );

  if ( loading ) return <p>Loading....</p>;
  if ( error ) return <ApolloError error={ error } />;
  if ( !data.videoProject ) return null;

  const {
    units,
    supportFiles,
  } = data.videoProject;

  const videoFiles = getVideoFiles( units );

  if ( getCount( videoFiles ) || getCount( supportFiles ) ) {
    return (
      <div className="details-files">
        <ul>
          { videoFiles && videoFiles.map( vidFile => {
            if ( !vidFile || getCount( vidFile ) === 0 ) return null;

            return (
              <li key={ vidFile.id }>
                { vidFile.use && vidFile.use.name }
                { ' | ' }
                { vidFile.quality }
                { ' | ' }
                { vidFile.language && vidFile.language.displayName }
                { ' ' }
                (
                { formatBytes( vidFile.filesize ) }
                )
              </li>
            );
          } ) }
          { supportFiles && supportFiles.map( sprtFile => {
            if ( !sprtFile || getCount( sprtFile ) === 0 ) return null;

            return (
              <li key={ sprtFile.id }>
                { 'SRT | ' }
                { sprtFile.language && sprtFile.language.displayName }
                { ' | ' }
                { formatBytes( sprtFile.filesize ) }
              </li>
            );
          } ) }
        </ul>
      </div>
    );
  }

  return <p>There are no supporting video files.</p>;
};

VideoDetailsPopup.propTypes = {
  id: PropTypes.string,
};

export default VideoDetailsPopup;

export { getValidFiles, getVideoFiles, VIDEO_PROJECT_FILES_QUERY };
