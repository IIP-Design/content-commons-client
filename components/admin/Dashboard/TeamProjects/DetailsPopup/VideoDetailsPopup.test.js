import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from '@apollo/client/testing';
import ApolloError from 'components/errors/ApolloError';
import { formatBytes, getCount, suppressActWarning } from 'lib/utils';
import VideoDetailsPopup, {
  getValidFiles, getVideoFiles, VIDEO_PROJECT_FILES_QUERY,
} from './VideoDetailsPopup';

jest.mock( 'next/config', () => () => ( {
  publicRuntimeConfig: { REACT_APP_AWS_S3_AUTHORING_BUCKET: 's3-bucket-url' },
} ) );

const props = { id: '123' };

const mocks = [
  {
    request: {
      query: VIDEO_PROJECT_FILES_QUERY,
      variables: { ...props },
    },
    result: {
      data: {
        videoProject: {
          id: props.id,
          projectType: 'LANGUAGE',
          units: [
            {
              id: 'u1914',
              files: [
                {
                  id: 'f281',
                  filesize: 6424220,
                  quality: 'BROADCAST',
                  url: `2019/06/${props.id}/video-1.mp4`,
                  language: {
                    id: 'en33',
                    displayName: 'English',
                  },
                  use: {
                    id: 'u382',
                    name: 'Full Video',
                  },
                },
                {
                  id: 'f294',
                  filesize: 6424220,
                  quality: 'BROADCAST',
                  url: `2019/06/${props.id}/video-2.mp4`,
                  language: {
                    id: 'fr533',
                    displayName: 'French',
                  },
                  use: {
                    id: 'u382',
                    name: 'Full Video',
                  },
                },
              ],
            },
          ],
          supportFiles: [
            {
              id: 'v832',
              url: `2019/06/${props.id}/srt-1.srt`,
              filesize: 6424,
              language: {
                id: 'en33',
                displayName: 'English',
              },
            },
            {
              id: 'v238',
              url: `2019/06/${props.id}/srt-2.srt`,
              filesize: 6424,
              language: {
                id: 'fr533',
                displayName: 'French',
              },
            },
          ],
        },
      },
    },
  },
];

const noFilesMocks = [
  {
    ...mocks[0],
    result: {
      data: {
        videoProject: {
          ...mocks[0].result.data.videoProject,
          units: [
            {
              ...mocks[0].result.data.videoProject.units[0],
              files: [],
            },
          ],
          supportFiles: [],
        },
      },
    },
  },
];

const noUnitsMocks = [
  {
    ...mocks[0],
    result: {
      data: {
        videoProject: {
          ...mocks[0].result.data.videoProject,
          units: [],
        },
      },
    },
  },
];

const nullFilesMocks = [
  {
    ...mocks[0],
    result: {
      data: {
        videoProject: {
          ...mocks[0].result.data.videoProject,
          units: [
            {
              ...mocks[0].result.data.videoProject.units[0],
              files: null,
            },
          ],
          supportFiles: null,
        },
      },
    },
  },
];

const nullFileMocks = [
  {
    ...mocks[0],
    result: {
      data: {
        videoProject: {
          ...mocks[0].result.data.videoProject,
          units: [
            {
              ...mocks[0].result.data.videoProject.units[0],
              files: [
                ...mocks[0].result.data.videoProject.units[0].files,
                null,
              ],
            },
          ],
          supportFiles: [
            ...mocks[0].result.data.videoProject.supportFiles,
            null,
          ],
        },
      },
    },
  },
];

const Component = (
  <MockedProvider mocks={ mocks } addTypename={ false }>
    <VideoDetailsPopup { ...props } />
  </MockedProvider>
);

describe( '<VideoDetailsPopup />', () => {
  const consoleError = console.error;

  beforeAll( () => suppressActWarning( consoleError ) );
  afterAll( () => {
    console.error = consoleError;
  } );

  it( 'renders initial loading state without crashing', () => {
    const wrapper = mount( Component );
    const popup = wrapper.find( 'VideoDetailsPopup' );
    const loading = <p>Loading....</p>;

    expect( popup.exists() ).toEqual( true );
    expect( popup.contains( loading ) ).toEqual( true );
  } );

  it( 'renders error message if an error is returned', async () => {
    const errorMocks = [
      {
        request: {
          query: VIDEO_PROJECT_FILES_QUERY,
          variables: { ...props },
        },
        result: {
          errors: [{ message: 'There was an error.' }],
        },
      },
    ];

    const wrapper = mount(
      <MockedProvider mocks={ errorMocks } addTypename={ false }>
        <VideoDetailsPopup { ...props } />
      </MockedProvider>,
    );

    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const popup = wrapper.find( VideoDetailsPopup );
    const errorComponent = popup.find( ApolloError );

    expect( errorComponent.exists() ).toEqual( true );
    expect( errorComponent.contains( 'There was an error.' ) )
      .toEqual( true );
  } );

  it( 'renders the final state', async () => {
    const wrapper = mount( Component );

    await wait( 0 );
    wrapper.update();

    const popup = wrapper.find( VideoDetailsPopup );

    expect( toJSON( popup ) ).toMatchSnapshot();
  } );

  it( 'renders null if videoProject is null', async () => {
    const nullMocks = [
      {
        request: {
          query: VIDEO_PROJECT_FILES_QUERY,
          variables: { ...props },
        },
        result: {
          data: { videoProject: null },
        },
      },
    ];
    const wrapper = mount(
      <MockedProvider mocks={ nullMocks } addTypename={ false }>
        <VideoDetailsPopup { ...props } />
      </MockedProvider>,
    );

    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const popup = wrapper.find( VideoDetailsPopup );

    expect( popup.html() ).toEqual( null );
  } );

  it( 'renders a no files message if there are no files', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ noFilesMocks } addTypename={ false }>
        <VideoDetailsPopup { ...props } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const popup = wrapper.find( VideoDetailsPopup );
    const noFilesMsg = <p>There are no supporting video files.</p>;

    expect( popup.contains( noFilesMsg ) ).toEqual( true );
  } );

  it( 'renders a no files message if files are null', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ nullFilesMocks } addTypename={ false }>
        <VideoDetailsPopup { ...props } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const popup = wrapper.find( VideoDetailsPopup );
    const noFilesMsg = <p>There are no supporting video files.</p>;

    expect( popup.contains( noFilesMsg ) ).toEqual( true );
  } );

  it( 'does not render video files if there are no units', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ noUnitsMocks } addTypename={ false }>
        <VideoDetailsPopup { ...props } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const popup = wrapper.find( VideoDetailsPopup );
    const files = popup.find( 'li' );
    const { videoProject: { supportFiles } } = noUnitsMocks[0].result.data;
    const supportFilesCount = getValidFiles( supportFiles ).length;

    expect( files.length ).toEqual( supportFilesCount );
    files.forEach( ( file, i ) => {
      const { language: { displayName }, filesize } = supportFiles[i];

      expect( file.text() )
        .toEqual( `SRT | ${displayName} | ${formatBytes( filesize )}` );
    } );
  } );

  it( 'renders the correct number of files', async () => {
    const wrapper = mount( Component );

    await wait( 0 );
    wrapper.update();

    const popup = wrapper.find( VideoDetailsPopup );
    const files = popup.find( 'li' );
    const { videoProject: { units, supportFiles } } = mocks[0].result.data;
    const supportFilesCount = getValidFiles( supportFiles ).length;
    const unitFiles = getVideoFiles( units );
    const unitFilesCount = getCount( unitFiles );
    const totalFiles = supportFilesCount + unitFilesCount;

    expect( files.length ).toEqual( totalFiles );
  } );

  it( 'does not crash and renders the correct number of files if any single file is null', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ nullFileMocks } addTypename={ false }>
        <VideoDetailsPopup { ...props } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const popup = wrapper.find( VideoDetailsPopup );
    const files = popup.find( 'li' );
    const { videoProject: { units, supportFiles } } = nullFileMocks[0].result.data;
    const supportFilesCount = getValidFiles( supportFiles ).length;
    const unitFiles = getVideoFiles( units );
    const unitFilesCount = getCount( unitFiles );
    const totalFiles = supportFilesCount + unitFilesCount;

    expect( files.length ).toEqual( totalFiles );
  } );
} );
