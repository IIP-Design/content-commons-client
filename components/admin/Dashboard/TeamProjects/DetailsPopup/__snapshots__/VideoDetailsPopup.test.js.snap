// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VideoDetailsPopup /> renders the final state 1`] = `
<VideoDetailsPopup
  id="123"
>
  <div
    className="details-files"
  >
    <ul>
      <li
        key="f281"
      >
        Full Video
         | 
        BROADCAST
         | 
        English
         
        (
        6.13  MB
        )
      </li>
      <li
        key="f294"
      >
        Full Video
         | 
        BROADCAST
         | 
        French
         
        (
        6.13  MB
        )
      </li>
      <li
        key="v832"
      >
        SRT | 
        English
         | 
        6.27  KB
      </li>
      <li
        key="v238"
      >
        SRT | 
        French
         | 
        6.27  KB
      </li>
    </ul>
  </div>
</VideoDetailsPopup>
`;
